pandas/core/generic.py

```diff
@@ -1948,6 +1948,129 @@ def tail(self, n=5):
+        """
+        Returns a random sample of rows from object. 
+        
+        Parameters
+        ----------
+        n : int, optional
+            Number of rows to return. Cannot be used with `frac`.
+            Default = 1 if `frac` = None. 
+        frac : float, optional
+            Share of rows to return. Cannot be used with `n`. 
+        replace : boolean, optional
+            Sample with or without replacement. Default = False. 
+        weights : str or ndarray-like, optional
+            Default 'None' results in equal probability weighting. 
+            If called on a DataFrame or Panel, will also accept the name of a 
+            column as a string. Must be same length as index. 
+            If weights do not sum to 1, they will be normalized to sum to 1. 
+            Missing values in the weights column will be treated as zero. 
+            inf and -inf values not allowed. 
+        random_state : int or numpy.random.RandomState, optional
+            Seed for the random number generator (if int), or numpy RandomState 
+            object.
+        axis : int, optional
+            Axis to sample. Default = 0.
+            
+        Returns
+        -------
+        Same type as caller. 
+        """
+        
+        ###        
+        # Processing axis argument
+        ###
+        
+        # This checks validity of axis argument.
+        if axis > (self.ndim - 1):
+            raise ValueError("Axis argument out of range")
+    
+        # Store length of relevant axis of object. 
+        axis_length = self.shape[axis]    
+    
+        ###
+        # Processing random_state argument
+        ###
+    
+        # random_state may be either a RandomState object, 
+        # integer, or None. Otherwise raise an error. 
+     
+        if isinstance(random_state, int):
+            rs = np.random.RandomState(random_state)
+        elif isinstance(random_state, np.random.RandomState):
+            rs = random_state
+        elif random_state is None:
+            rs = np.random.RandomState()
+        else:
+            raise ValueError("Seed must be either an integer or numpy RandomState")
+    
+        ###
+        #  Process weight argument
+        ###
+    
+        # Check weights for compliance
+        if weights is not None:
+                                      
+            # Strings acceptable if not a series
+            if isinstance(weights, string_types): 
+
+                if self.ndim > 1 :
+                    try:
+                        weights = self[weights]
+                    except KeyError:
+                        raise KeyError("String passed to weights not a valid column name")
+
+                else:
+                    raise ValueError("Strings cannot be passed as weights when sampling from a Series.")
+
+            #normalize format of weights to ndarray. 
+            weights = np.asarray(weights, dtype = np.float)
```

Comment:
who say it doesn't accept a Series.
Try to use pandas functions as much as possible. Eg. then you don't have to use odd things like ``np.nan_to_num`` (which doesn't work for most dtypes)
```
In [88]: np.random.choice(Series([1,2,3]))
Out[88]: 2
```